iml-services/iml-snapshot/src/retention.rs

```diff
@@ -0,0 +1,213 @@
+            let percent_free = 100.0f64 - percent_used;
+
+            tracing::debug!(
+                "stats record: {:?} - bytes free: {}",
+                stats_record.get(&fs_name),
+                bytes_free
+            );
+            let retention = get_retention_policy(pool, &fs_name).await?;
+
+            if let Some(retention) = retention {
+                let snapshots = get_snapshots(pool, &fs_name).await?;
+
+                tracing::debug!(
+                    "percent_left: {}, reserve value: {}",
+                    percent_free,
+                    retention.reserve_value
+                );
+                let should_delete_snapshot = match retention.reserve_unit {
+                    snapshot::ReserveUnit::Percent => percent_free < retention.reserve_value as f64,
+                    snapshot::ReserveUnit::Gibibytes => {
+                        let gib_free: f64 = bytes_free as f64 / 1_073_741_824_f64;
+                        gib_free < retention.reserve_value as f64
+                    }
+                    snapshot::ReserveUnit::Tebibytes => {
+                        let teb_free: f64 = bytes_free as f64 / 1_099_511_627_776_f64;
+                        teb_free < retention.reserve_value as f64
+                    }
+                };
+                tracing::debug!("Should delete snapshot?: {}", should_delete_snapshot);
+
+                if should_delete_snapshot
+                    && snapshots.len() > retention.keep_num as usize
+                    && (stats_record.get(&fs_name).is_none()
+                        || stats_record.get(&fs_name) != Some(&bytes_used))
+                {
+                    stats_record.insert(fs_name.to_string(), bytes_used);
+                    tracing::debug!("About to delete earliest snapshot.");
+                    let snapshot_name = snapshots[0].snapshot_name.to_string();
+                    tracing::debug!("Deleting {}", snapshot_name);
+                    let cmd =
+                        destroy_snapshot(client.clone(), &fs_name, snapshot_name.as_ref()).await?;
+
+                    wait_for_cmds_success(&vec![cmd], None).await?;
+                }
+            }
+        }
+    }
+
+    Ok(stats_record)
+}
+
+pub async fn handle_retention_rules(
+    client: Client,
+    influx_client: InfluxClient,
+    pool: PgPool,
+) -> Result<(), Error> {
+    let mut prev_stats: HashMap<String, u64> = vec![].into_iter().collect::<HashMap<String, u64>>();
+
+    loop {
+        prev_stats =
+            match process_retention(&client, &influx_client, &pool, prev_stats.clone()).await {
+                Ok(x) => x,
+                Err(e) => {
+                    tracing::error!("Retention Rule processing error: {:?}", e);
+                    prev_stats
+                }
+            };
+
+        // TODO change this to 60 seconds. 10 is for debugging.
+        tokio::time::delay_for(tokio::time::Duration::from_secs(30)).await;
```

Comment:
```suggestion
        tokio::time::delay_for(tokio::time::Duration::from_secs(60)).await;
```

?

Reply:
I had upped to 30 but yeah we probably want 60. I'll update.